### DISCUSSION TITLE
Allow configuration entities to store settings locally for other entities

### DISCUSSION THREAD

**Discussion by epenet** [home-assistant/architecture#713] [Entity Models]
It would be great if we could use an Entity (category=EntityCategory.CONFIG) to set some configuration settings for another Entity.

For example, we could use a SelectEntity to read/write a value to the config_entry data, and a SensorEntity could read the value to choose the correct endpoint to hit. This was rejected on discord, but maybe there is a now a better place to store this information?

Could the new entity options be used for such a thing? (draft PR https://github.com/home-assistant/core/pull/64350)

For example, in the  `onewire` integration, the `DS2760 1-Wire TypeK Thermocouple` provides eight endpoints to retrieve a temperature (one endpoint per thermocouple types: types B, E, J, K, N, R, S and T).
Having a SelectEntity to select the type, and writing that value into the registry options of the SensorEntity would avoid creating 8 disabled sensors.

**Comment by frenck** [comment]
The term you are looking for, is device options.

We currently do not allow for virtual entities.
This has recently been re-evaluated and decided not to add such a future in the short term.

The linked PR above are entity settings for entities themselves (not for other entities).

**Reply by epenet** [reply]
Where should these device options be stored? Is `config_entry.options` the best location?

**Reply by frenck** [reply]
We don't have a construct for device-based options in Home Assistant at this point.

**Comment by tetienne** [comment]
Well it seems related to this PR: https://github.com/home-assistant/core/pull/64350

**Reply by frenck** [reply]
@tetienne No, it is not. IMHO.
Additionally, that was already linked in the OP.

**Reply by tetienne** [reply]
From my understanding, this PR allow to add option to an entity and I guess it’s what @epenet wish.

**Reply by epenet** [reply]
That was my original thought also, and I mentionned it my original description.
But who/what is able to write into this if not virtual entities? Should a config_flow write to it?

It seems for now it will be restricted to a specific set hardcoded in the frontend "like overriding the unit and precision for a sensor, max volume for a media player and maybe a few other things"

**Reply by Jc2k** [reply]
As I read it, this is about reusing the new config entity types (that let you expose setting s **on the device itself** as entities) to expose settings on a device page potentially with that PR as a backend. But that PR is irrelevant to the discussion IMHO.

**Reply by Jc2k** [reply]
Also the UX described here would create device options. The linked PR is for entity options.

If you create a device level option, how do you know what entity it is for?

**Reply by frenck** [reply]
> But who/what is able to write into this if not virtual entities? Should a config_flow write to it?

Entities are only allowed to expose data points provided by the device itself.

**Comment by Jc2k** [comment]
I think device options could be shown on the devices page like these entities are, but it seems a bit weird to make fake entities to do that.

Could we associate a config flow like schema with a device and use that to create extra widgets on the device page?

**Comment by epenet** [comment]
In the case of the DS2760, the device exposes 3 data points per thermocouple type, so 24 data points :
- `{device_id}/typeB/temperature`
- `{device_id}/typeB/ranges_low`
- `{device_id}/typeB/ranges_high`
- ...
- `{device_id}/typeT/temperature`
- `{device_id}/typeT/ranges_low`
- `{device_id}/typeT/ranges_high`

We want to ensure only 3 entities per device instead of 24 (19 disabled, 3 enabled), and there are two main questions that need to be answered.

The first question is where to store the setting:
option 1 => we store the setting in the entity options, and the setting has to be written three times (could be automated by a clever UI)
option 2 => we store the setting in new device options, and the setting is written just once for all three entities
option 3 => we store the setting in the config entry options, and the setting is written just once for all three entities
option 4 => ???

The second question is where does the UI fit:
option a => ~virtual entity~ (currently not allowed)
option b => standard widget in the frontend, passing in a schema
option c => custom widget in the frontend
option d => options flow in the config-entry
option e => ???

**Reply by frenck** [reply]
options flow is the only method available for such things at this point. But however, considering the goal:

> We want to ensure only 3 entities per device instead of 24 (19 disabled, 3 enabled), and there are two main questions that need to be answered.

The options flow should not handle enabling/disabling of entities at all.

**Reply by Jc2k** [reply]
Out of interest, why are 19 disabled? Can you "probe the probe" and work out which one to use?

**Reply by epenet** [reply]
> Out of interest, why are 19 disabled? Can you "probe the probe" and work out which one to use?

Sadly "probe the probe" is not a possibility here.

**Reply by epenet** [reply]
> The options flow should not handle enabling/disabling of entities at all.

I agree. The options flow will not enable/disable any entity. It will just write the selected type in the settings.
Then during the sensor loading process only 3 entities will be created with the correct type:
- `{device_id}/type/temperature` with endpoint `{device_id}/type{thermo_type}/temperature`
- `{device_id}/type/ranges_low` with endpoint `{device_id}/type{thermo_type}/ranges_low`
- `{device_id}/type/ranges_high` with endpoint `{device_id}/type{thermo_type}/ranges_high`

**Reply by MartinHjelmare** [reply]
We do allow an options flow for disabling entities if there's a clear group of entities being disabled instead of just disabling non related entities.

https://developers.home-assistant.io/docs/entity_registry_disabled_by#integrations-offering-options-to-control-disabled_by

**Reply by droberg** [reply]
Just to clarify, in the case of the DS2760, the underlying service does indeed expose 3 data points per thermocouple type, so 24 in total using this way of putting it. However, only one thermocouple type is valid at any given point. So, having 24 entities while 19 is disabled and 3 enables is a weird thing to to. You can think of it as selecting between a bunch of calibration constants.

**Reply by MartinHjelmare** [reply]
As noted in the docs link above, when integrations offer an options flow to "disable" a group of entities, those "disabled" entities should be removed from the device and entity registry completely. So there won't be any entities left from that group marked as disabled.
